src/equisolve/numpy/models/linear_model.py

```diff
@@ -147,17 +165,70 @@ def _solver(
         sw_arr = sw_arr.ravel()
         alpha_arr = alpha_arr.ravel()
 
-        # Convert problem with regularization term into an equivalent
-        # problem without the regularization term
         num_properties = X_arr.shape[1]
 
-        regularization_all = np.hstack((sw_arr, alpha_arr))
-        regularization_eff = np.diag(np.sqrt(regularization_all))
-
-        X_eff = regularization_eff @ np.vstack((X_arr, np.eye(num_properties)))
-        y_eff = regularization_eff @ np.hstack((y_arr, np.zeros(num_properties)))
-
-        w = np.linalg.lstsq(X_eff, y_eff, rcond=rcond)[0]
+        sw_arr = sw_arr.reshape(-1, 1)
+        sqrt_sw_arr = np.sqrt(sw_arr)
+
+        if self._solver == "auto":
+            n_samples, n_features = X_arr.shape
+
+            if n_features > n_samples:
+                # solves linear system (K*sw + alpha*Id) @ dual_w = y*sw
+                dual_alpha_arr = X_arr @ alpha_arr
+                X_eff = sqrt_sw_arr * X_arr
+                y_eff = y_arr * sw_arr.flatten()
+                K = X_eff @ X_eff.T + np.diag(dual_alpha_arr)
+                try:
+                    dual_w = scipy.linalg.solve(K, y_eff, assume_a="pos").ravel()
+                except np.linalg.LinAlgError:
```

Comment:
Should we raise a warning that we fallback to `lstsq`?

Reply:
I had this first, but then I decided to rename the solver to "auto" and make it clear in the doc that it switches solver if one fails. People who don't care, won't get bothered with warnings. People who care will read this and understand. If these people want to use an explicit solver they change the argument. For explicit solver arguments we do not change the solver.

But I am also fine to add an argument `raise_warning=False` that when enabled, raises warnings once the solver changes

Reply:
No it is okay if we use auto that everything is done automatic